test/widgets/test_windowtabs.py

```diff
@@ -0,0 +1,137 @@
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# Widget specific tests
+
+import pytest
+
+import libqtile.config
+from libqtile import bar, layout, widget
+from libqtile.config import Screen
+from libqtile.confreader import Config
+
+
+class FakeScreenConfig(Config):
+    auto_fullscreen = True
+    groups = [
+        libqtile.config.Group("a"),
+        libqtile.config.Group("b")
+    ]
+    layouts = [
+        layout.Stack()
+    ]
+    floating_layout = libqtile.resources.default_config.floating_layout
+    keys = []
+    mouse = []
+    fake_screens = [
+        Screen(
+            top=bar.Bar(
+                [
+                    widget.WindowTabs(),
+                ],
+                24,
+            ),
+            bottom=bar.Bar(
+                [
+                    widget.WindowTabs(selected="!!"),
+                ],
+                24,
+            ),
+            x=0, y=0, width=900, height=960,
+        ),
+    ]
+    screens = []
+
+
+xephyr_config = {
+    "xinerama": False,
+    "two_screens": False,
+    "width": 900,
+    "height": 960
+}
+fakescreen_config = pytest.mark.parametrize("xephyr, manager", [(xephyr_config, FakeScreenConfig)], indirect=True)
+
+
+@fakescreen_config
+def test_single_window_states(manager):
+
+    def widget_text():
+        return manager.c.bar["top"].info()["widgets"][0]["text"]
+
+    # Default _TextBox text is " " and no hooks fired yet.
+    assert widget_text() == " "
+
+    # Load xeyes
+    proc = manager.test_xeyes()
+    assert widget_text() == "<b>xeyes</b>"
+
+    # Maximize window
+    manager.c.window.toggle_maximize()
+    assert widget_text() == "<b>[] xeyes</b>"
+
+    # Minimize window
+    manager.c.window.toggle_minimize()
+    assert widget_text() == "<b>_ xeyes</b>"
+
+    # Float window
+    manager.c.window.toggle_minimize()
+    manager.c.window.toggle_floating()
+    assert widget_text() == "<b>V xeyes</b>"
+
+    # Kill xeyes and check text again
+    # NB hooks fired so empty string is now ""
+    manager.kill_window(proc)
+    assert widget_text() == ""
+
+
+@fakescreen_config
+def test_multiple_windows(manager):
+
+    def widget_text():
+        return manager.c.bar["top"].info()["widgets"][0]["text"]
+
+    window_one = manager.test_window("one")
+    assert widget_text() == "<b>one</b>"
+
+    window_two = manager.test_window("two")
+    assert widget_text() in ["<b>two</b> | one", "one | <b>two</b>"]
```

Comment:
I did it this way as the order seemed to change randomly so test would pass some times but not others. Is it ok to do this way or should the order be fixed?

Reply:
Yeah I think we should have this fixed. :)

Reply:
OK. May leave this as is for now because it's not the widget's fault. I'd need to do some more digging to work out why this is happening.

Reply:
As above - will take a look separately and can submit a PR if it needs fixing.